### PR TITLE
Benchmarking and iterator pattern

### PR DESCRIPTION
**Pull request by attackgoat**
I took a stab at improving the performance of this library.
- Added benchmarks for parse operation
- Converted `Vec` return types to `Iterator` implementations
- Added some documentation
- Changed "parse errors" to be a Cargo feature; I wanted different behavior for my use cases plus this sped things up a bit
- Changed to Rust 2018 edition because of some tooling issue; could probably resolve that but it is 2021 anyways

Because of the new cargo feature it may be helpful to run both `cargo test` and also `cargo test --no-default-features` in order to run the tests for both paths.

To run the benchmarks, run `cargo bench` from the root, or just cherry-pick `e9c43f1` onto some other branch you might want to bench. Each of the things on this branch is separable (with some work) and individually improves performance.

## Sample
- `Winner: Attackgoat`: Before=**2.9**us After=**112**ns
- `Van Gogh my earlo...[~520 CHARS]`: Before=**82**us After=**3.6**us

This is the perf graph for `Winner: Attackgoat`; red is this PR and blue is the `master` branch with `e9c43f1`:

image

## TL;DR:

Criterion says this is -95% improvement.

### CODE REVIEW

**Comment by KalitaAlexey** [comment]
omg :)

File: benches/parse.rs [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    criterion::{black_box, criterion_group, criterion_main, BenchmarkId, Criterion},
+    std::fs::File,
+};
+
+const INPUTS: &[(&str, &str)] = &[
+    ("01: Empty", ""),
+    ("02: Character", "w"),
+    ("03: Word", "Score"),
+    ("04: Pair", "Winner: Attackgoat"),
+    ("05: Sentance", "The lazy dog can sit wherever she wishes."),
```
Comment (KalitaAlexey) [inline benches/parse.rs]: Sentence here and below.

File: src/lib.rs [hunk 2]
```diff
@@ -45,6 +72,34 @@ pub struct BMFont {
...
 
 impl BMFont {
+    /// Constructs a new [BMFont].
+    ///
+    /// ## Examples
+    ///
+    /// From a file:
+    ///
+    /// ```rust
+    /// # use bmfont::*;
```
Comment (KalitaAlexey) [inline src/lib.rs:82]: Why all these comments? Here and below.

File: src/lib.rs [hunk 3]
```diff
@@ -45,6 +72,34 @@ pub struct BMFont {
 }
 
 impl BMFont {
+    /// Constructs a new [BMFont].
+    ///
+    /// ## Examples
```
Comment (KalitaAlexey) [inline src/lib.rs]: I think it should be ###, not sure though.

Comment (attackgoat) [reply src/lib.rs]: I was using ## from some other code base as an example; but now that I look at [the Rust source](https://doc.rust-lang.org/src/core/iter/traits/iterator.rs.html#111) it looks like # might be better in this situation.

I'll change it to that - feel free to modify/say or we let someone else provide substantial docs?

Comment (KalitaAlexey) [reply src/lib.rs]: Agree with #.

File: src/lib.rs (hunk 2 above)
Comment (attackgoat) [reply src/lib.rs:82]: I was following [this guidance](https://doc.rust-lang.org/rustdoc/what-to-include.html) about hiding boilerplate stuff.

In my library I've put a really detailed doc example at the library root in the ///! section and then I'm trying to make the other examples not show that stuff. YMMV, this depends on how you'd like to document this crate.

In the larger sense, the comments are there because this code gets compiled and tested with the `cargo test` command.
